### ISSUE TITLE
gRPC slower than REST for higher number of `repeated` elements

### ISSUE THREAD

**Issue by ashj11** [grpc/grpc-java#10891]
### What version of gRPC-Java are you using?
Using gRPC via https://github.com/grpc-ecosystem/grpc-spring. The grpc-java version used by that module is 1.58.0( https://github.com/grpc-ecosystem/grpc-spring/blob/a2e9520b2414277038a5a55551529f5023e8520b/docs/en/server/getting-started.md?plain=1#L52 )

### What is your environment?
Reproducible in multiple environments. Have tried on my Macbook, as well as an EC2 with Ubuntu. 
Java 17
Client and server are on two different EC2 machines in the same VPC in the same AWS region.

### What did you expect to see?
I expected gRPC to be faster even if  if the response returned has high number of repeated elements.

### What did you see instead?
gRPC was slower than REST with json when `repeated` elements are more than around 25. Below is the graph showing the trend. X axis shows the response time( p95 ) in ms. 
<img width="641" alt="image" src="https://github.com/grpc/grpc-java/assets/48014387/40965d22-3146-4508-9b8f-c2b9297a64e9">

### Steps to reproduce the bug

**Test Setup** 
Code - https://github.com/ashj11/grpc-rest-test
This is a simple springboot project which serves both gRPC and REST endpoints. the server just returns a precreated json/proto object when the method is called. 
[Grpc-Client-Page-children.py](https://github.com/ashj11/grpc-rest-test/blob/main/Grpc-Client-Page-children.py) is the client which calls both gRPC and REST endpoints for 100 iterations, for every child count in 1 to 100. The requests are made sequentially. 
gRPC service - [MyGRPCUpstreamImpl.java](https://github.com/ashj11/grpc-rest-test/blob/main/my-upstream-service/src/main/java/service/MyGRPCUpstreamImpl.java)
REST service - [MyRestUpstreamImpl.java](https://github.com/ashj11/grpc-rest-test/blob/main/my-upstream-service/src/main/java/service/MyRestUpstreamImpl.java)
Proto definition - [my_upstream.proto](https://github.com/ashj11/grpc-rest-test/blob/main/my-upstream-interface/src/main/proto/my_upstream.proto)
Found a [Slack overflow link](https://stackoverflow.com/questions/69889439/why-is-grpc-so-much-slower-than-an-http-api-sending-an-array) which stats this is expected. But I was surprised to see the results and would like to confirm if gRPC is not expected to handle repeated elements more than say 25.

**Comment by ejona86** [comment]
We would need to play to confirm, but I doubt grpc-java has much to do with the performance you are seeing. `message Child` looks relatively small; small enough that I'd expect 100 repetitions to be within a ms (on a warm JVM).

The most likely explanation is the python code is using a python-powered protobuf. Protobuf in Python has two forms: python-based and c-based. The `json` module in Python is a c module, so you probably need to use the c-based protobuf for apples-to-apples. The main protobuf project provided both forms, and I thought the newer generated code used the c++ decoder, but for the specifics for checking/changing which you are using I am not much help.

**Comment by ashj11** [comment]
Added a small note that client and server are different EC2 machines on same VPC in same AWS region. 

Thanks for you reply @ejona86 . Verified on client that the protobuf implementation used by python is [ubp](https://github.com/protocolbuffers/upb), which is the C implementation. Command used `python3 -c "from google.protobuf.internal import api_implementation; print(api_implementation.Type())"`
Also in the [stackoverflow](https://stackoverflow.com/questions/69889439/why-is-grpc-so-much-slower-than-an-http-api-sending-an-array) link that i have mentioned, the implementation is in Node.JS. 
It is possible this is not an issue with the java implementation per se, but with protobufs in general as mentioned in the answer there(?)

**Comment by ejona86** [comment]
The stackoverflow question is adding MBs of repeated integers, which is _a lot_ more than involved here. And Node.js would naturally favor json more than Python would, since it is pure-JS. Thanks for confirming the protobuf implementation you were testing on.

**Comment by ashj11** [comment]
Also uploaded [my-upstream-service-1.0-SNAPSHOT.jar](https://github.com/ashj11/grpc-rest-test/blob/main/my-upstream-service-1.0-SNAPSHOT.jar) so that you can try running the service with `java -jar` command.

**Comment by ashj11** [comment]
Any update on this?

**Comment by ejona86** [comment]
I think a big source of slow-down in your results are a cold JVM. Since the server was hard-coded to a single response size, you restarted the server after each one and I expect you didn't run long enough to let the server warm up. To be fully warm, you need 30+ seconds of continuous execution, but even after just a few seconds you'll see the numbers change dramatically.

Note that it is a bit important that you are using grpc_requests: https://pypi.org/project/grpc-requests/ . That isn't the normal client, and so may not behave the same when using code generation. But given what I know about how Protobuf parses messages in Python normally, it might be possible it doesn't matter much, even if it is going through different code paths for some "hot" code. Maybe it matters, maybe it doesn't, but it does mean the results are murkier.

I started a server with `NUM_OF_CHILDREN = 50` and had the python `numberOfIterations = 100`. From a cold JVM and running the python client multiple times:
```
# 1st run
Rest p95 : 2.0098499999999992
GRPC p95 : 0.9072999999999998
# 2nd
Rest p95 : 1.8455999999999997
GRPC p95 : 0.51625
# 3rd
Rest p95 : 1.2263499999999998
GRPC p95 : 0.4451499999999999
# 15th
Rest p95 : 0.8654499999999999
GRPC p95 : 0.2571
```

You can see how important the warm-up is for both REST and gRPC. Even with doing `numberOfIterations = 1` I saw REST as slower. That may be because I'm doing this on a beefy desktop and not a laptop. But more importantly is doing the run many times. For very few iterations the first thing executed is typically slower because it needs to do slow class loading. But I wasn't able to make that explain our different results.

`NUM_OF_CHILDREN = 10000`, `numberOfIterations = 1`:
```
# grpc first
('grpc.health.v1.Health', 'grpc.reflection.v1alpha.ServerReflection', 'myupstream.MyUpstream')
GRPC iteration 0 completed
GRPC p95 : 1.407
Rest iteration 0 completed
Rest p95 : 89.52

# I restarted the server here, then ran again.

# rest first
Rest iteration 0 completed
Rest p95 : 86.277
('grpc.health.v1.Health', 'grpc.reflection.v1alpha.ServerReflection', 'myupstream.MyUpstream')
GRPC iteration 0 completed
GRPC p95 : 1.1199999999999999
```

That's about as mean I can be with gRPC, because gRPC tends to get much faster with more iterations. But REST was hit particularly bad by class loading/cold JVM. Dunno.

Now for something more useful. `NUM_OF_CHILDREN = 10000`, `numberOfIterations = 100`:
```
# 1st
Rest p95 : 1.9756499999999995
GRPC p95 : 1.0613500000000002
# 2nd
Rest p95 : 1.9456999999999998
GRPC p95 : 0.50545
# 3rd
Rest p95 : 1.7209999999999999
GRPC p95 : 0.42214999999999997
# 15th
Rest p95 : 1.6706499999999997
GRPC p95 : 0.27829999999999994
```

**Comment by ejona86** [comment]
(You probably saw an earlier response in your email, now deleted. I was accidentally testing with `Grpc-Client.py`, so it was sort of useless. This was done with a hacked `Grpc-Client.py` that was changed to call GetPage/pageinfo; I got some errors I didn't want to bother with on the other one. I also set NUM_OF_CHILDREN = 10000 when building the server)

The slowdown definitely seems to be on the Python side. I am worried about slow JVM warmup, but it is minor here compared to the slowdown on the client.

I see you are using [grpc_requests](https://pypi.org/project/grpc-requests/). That is not the normal client, and it may be the cause of the slow-down here. It might not cause a slow-down since even the normal Python processing uses reflection information, but it does mean it isn't worth digging in too deeply.

`NUM_OF_CHILDREN = 10000` with full Python-side decoding (not fair):
```
Rest p95 : 44.494299999999996
GRPC p95 : 228.44089999999997
```

`NUM_OF_CHILDREN = 10000` using `client.request(service, method, {}, raw_output=True)` to disable grpc_requests decoding (fair):
```
Rest p95 : 43.82405
GRPC p95 : 19.9179
```

Since the server responded in ~20 ms, that means ~200 ms of latency was on the Python side. This proves the problem isn't with Java. Also, this comparison is fair, since requests isn't decoding the JSON either.

To compare them both decoding, `NUM_OF_CHILDREN = 10000` with original grpc decoding and adding a  `requests.json()` for REST (fair):
```
Rest p95 : 51.73215
GRPC p95 : 227.7057
```

I don't think there's any more to do here from the Java side.
